### ISSUE TITLE
HttpRequest "normalizing" headers assumes all callee's are case insensitive

### ISSUE THREAD

**Issue by i8beef** [node-red/node-red#3081]
### Current Behavior

When setting headers for an HttpRequest, it appears to "normalize" all of them to lowercase, which means requests to endpoints that are actually case sensitive have no avenue to actually function.

### Expected Behavior

Headers should be piped through as specified for all headers rather than transforming them.

### Steps To Reproduce

Send a request to any device / endpoint that makes a case-sensitive header evaluation (In my case, it was the "Host" header always being translated to "host", and the callee not liking that.

### Example flow

```
var newMsg = {};
newMsg.url = "endpointThatCaresAboutHostHeaderCase";
newMsg.headers = {};
newMsg.headers["Host"] = "192.168.1.1";
return newMsg;
```

### Environment

- Node-RED version: 2.0.2
- Node.js version: Docker
- npm version: Docker
- Platform/OS: Docker
- Browser: Chrome

**Comment by i8beef** [comment]
(One of) Location where I believe the "header normalization" / "toLowerCase" is occuring: https://github.com/node-red/node-red/blob/master/packages/node_modules/%40node-red/nodes/core/network/21-httprequest.js#L200

**Comment by hardillb** [comment]
The HTTP 1.1 spec says that header names should not be case sensitive
https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2

> Each header field consists of a name followed by a colon (":") and the field value. Field names are case-insensitive.

(Not saying we won't look at this, but your end point is not to spec if it's case sensitive)

**Comment by knolleary** [comment]
I know the old request module went to great lengths to be consistent with the headers it set. Looks like the new module we use doesn't take the same care.

Spec aside, this is a regression in behaviour from 1.x so we need to do something.

**Comment by i8beef** [comment]
Unfortunately, not my endpoint so I can only control one side of the equation here. Don't disagree with you, and I'll try reporting this to the device manufacturer, but its one of those standalone basic devices that I'm guessing me complaining about them being out of HTTP spec is gonna get summarily ignored 😄

**Comment by hardillb** [comment]
Looking at the code that line shouldn't be changing the case unless it's `content-type` or `content-length` (unless I'm miss reading the if statement under it)

But I have wiresharked the actual request and it looks like got (or one of the agents) is normalising all the headers to lowercase

**Comment by hardillb** [comment]
This looks promising - https://github.com/sindresorhus/got/pull/1336

**Comment by knolleary** [comment]
@hardillb sadly not - its closed without being merged. If you review all the linked issues, they standard response is "headers should be case insensitive - fix your server"

In one, they say "maybe you should be using the `request` module if thats what you want"... oh the irony.

It looks like this is what we need - https://github.com/sindresorhus/got/pull/1103#issuecomment-620364057 - using their hooks api to 'fix' the headers before it sends the request.

**Comment by knolleary** [comment]
Whilst writing some unit tests for a different HTTP Request issue, I spotted we have a unit test that checks headers are always sent lower-case. That test is from 1.x, so I just checked our 1.x behaviour and it is the same - headers are lower-cased.

I was under the impression this was a regression in behaviour, but maybe not.

@i8beef can you confirm whether this is something that was working in 1.x for you, or are you only just trying to do this with 2.x?

**Comment by i8beef** [comment]
Yes this worked in 1.0, nothing else has changed except the upgrade to 2.x...

The only symptom I have really is that, and the error:

```
HTTPError: Response code 400 (Bad Request: No Host Header)
    at Request.<anonymous> (/usr/src/node-red/node_modules/got/dist/source/as-promise/index.js:117:42)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

I fired up a .NET project to make the same request... but .NET has a series of headers (Host among them) that are "special" and always sent the same way... so I can't actually reproduce there because Microsoft won't LET me send a "host" header, it will ALWAYS be "Host". That's rather interesting though I have no idea if its due to any sort of standard for those since that particular heads is part of the main HTTP 1.1 spec...

Thats my best educated GUESS as to what changed that is breaking, but this is the core SYMPTOM I could be incorrectly disagnosing...

**Comment by i8beef** [comment]
I played with a `curl` a bit with this device... I can't actually figure out how to change the 'Host' header (how curl sends it) since its automatic there... but I CAN change the "Authorization" header to be lowercase and said device fails to recognize the authentication anymore.

It also appears that if I don't inject a new `headers` object, but instead apply the Authorization in the node itself, two things change: (1) It doesn't complain about the Host header anymore (assuming that deep in the bowls of this `got` library, its actually auto generating the `Host` header TitleCased if left to its own devices?), but (2) the basic auth of course doesn't work because THAT is being changed to "authorization" instead of "Authorization".

I.e., I think I confirmed the device in question is in fact case sensitive here (incorrectly)... but also that other large frameworks, `curl`, etc. automatically generate TitleCased headers for these common use headers... interesting data point if nothing else about the "just lowercase" it community... I would argue they are PRACTICALLY wrong even if TECHNICALLY right 😄 

Note I can probably run this through my proxy server to fix this if you find you don't want to change this (or can't), no hard feelings.

**Comment by knolleary** [comment]
> Whilst writing some unit tests for a different HTTP Request issue, I spotted we have a unit test that checks headers are always sent lower-case. That test is from 1.x, so I just checked our 1.x behaviour and it is the same - headers are lower-cased.

Looking again, I was wrong on this analysis. The way the tests worked, they were seeing lower-cased headers because thats what the Express server used in the test does to normalise the received headers. When I looked at the `rawHeaders` object, I could see the request was sending headers with unmodified case.

I've pushed a fix that will restore the case of any headers the user has set - and fixed the test case to properly verify that works.
